lms/static/scripts/frontend_apps/utils/jwt.js

```diff
@@ -0,0 +1,54 @@
+/**
+ * @typedef JWTPayload
+ * @prop {number} exp
+ * @prop {number} nbf
+ */
+
+/**
+ * Value class for working with JSON Web Tokens issued by the server.
+ */
+export class JWT {
+  /**
+   * Construct a JWT to wrap a recently issued JWT token.
+   *
+   * @param {string} token - Serialized JWT
+   * @param {number} issuedAt -
+   *   A _client_ timestamp in milliseconds estimating when the JWT was issued.
+   *   The estimate should bias towards being earlier than the true time, in
+   *   which case the JWT will "expire" earlier than the true expiry.
+   */
+  constructor(token, issuedAt) {
+    this._token = token;
+
+    const [, payloadBase64] = token.split('.');
+    this._payload = /** @type {JWTPayload} */ (JSON.parse(atob(payloadBase64)));
+
+    // Estimated offset of server's clock relative to client.
+    //
+    // +ve if the server's clock is ahead of the client or -ve if the server's
+    // clock is behind.
+    const skew = this._payload.nbf * 1000 - issuedAt;
+
+    this._validUntil = this._payload.exp * 1000 - skew;
+  }
+
+  /**
+   * Return true if the JWT token has expired.
+   *
+   * @param {number} [now] - Current timestamp in milliseconds
+   */
+  hasExpired(now = Date.now()) {
+    return now > this._validUntil;
+  }
+
+  value() {
+    if (this.hasExpired()) {
+      throw new Error('Tried to use an expired JWT token');
+    }
```

Comment:
I think the responsibility of the backend server to validate the token, including the expire date.

Reply:
The h server does validate the token, but only once the client has received it and submitted it to the backend. The purpose of this code is to quickly check whether a refresh is needed before sending the token on to the client or for checking if an authorization failure may have been specifically caused by token expiry.

I'll update the `hasExpired` docs to clarify this.